File: example-go/src/github.com/cam-inc/dmc/example-go/controller/dmc.go
package controller

import (
	"encoding/json"

	"strings"

	"github.com/cam-inc/dmc/example-go/bridge"
	"github.com/cam-inc/dmc/example-go/common"
	"github.com/cam-inc/dmc/example-go/gen/app"
	jwtgo "github.com/dgrijalva/jwt-go"
	"github.com/goadesign/goa"
)

// DmcController implements the dmc resource.
type DmcController struct {
	*goa.Controller
}

func apiFilter(pages []*app.Page, roles map[string][]string) []*app.Page {
	newPages := []*app.Page{}
	for _, page := range pages {
		newComponents := []*app.Component{}
		for _, component := range page.Components {
			api := component.API
			method := api.Method
			resource := strings.Split(api.Path, "/")[1]
			if roles[method] != nil && (common.InStringArray("*", roles[method]) >= 0 || common.InStringArray(resource, roles[method]) >= 0) {
				newComponents = append(newComponents, component)
			}
		}
		if len(newComponents) > 0 {
			// componentが1つもない場合はpageも返さない
			page.Components = newComponents
			newPages = append(newPages, page)
		}
	}
	return newPages
}

// NewDmcController creates a dmc controller.
func NewDmcController(service *goa.Service) *DmcController {
	return &DmcController{Controller: service.NewController("DmcController")}
}

// Show runs the show action.
func (c *DmcController) Show(ctx *app.ShowDmcContext) error {
	// DmcController_Show: start_implement

	// Put your logic here
	defaultOptions := []*app.Option{
		{
			Key:   "key",
			Value: "value",
		},
	}

	api := func(path string, method string) *app.API {
		return &app.API{
			Path:   path,
			Method: method,
		}
	}

	// DmcController_Show: end_implement
	res := &app.Dmc{
		Name:      "Example Project",
		Color:     bridge.ColorRed,
		Thumbnail: "https://avatars3.githubusercontent.com/u/23251378?v=3&s=200",
		Tags: []string{
			"develop",
			"dmc",
			"example",
		},
		Pages: []*app.Page{
			// Dashboard
			{
				ID:      "quickview",
				Name:    "クイックビュー",
				Section: bridge.SectionDashboard,
				Group:   bridge.GroupEmpty,
				Components: []*app.Component{
					{
						Name:       "DAU",
						API:        api("/stats/dau", "get"),
						Style:      bridge.StyleNumber,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "MAU",
						API:        api("/stats/mau", "get"),
						Style:      bridge.StyleNumber,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(bar)",
						API:        api("/stats/planet/bar", "get"),
						Style:      bridge.StyleGraphBar,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(scatterplot)",
						API:        api("/stats/planet/scatterplot", "get"),
						Style:      bridge.StyleGraphScatterplot,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(line)",
						API:        api("/stats/planet/line", "get"),
						Style:      bridge.StyleGraphLine,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(horizontal-bar)",
						API:        api("/stats/planet/horizontal-bar", "get"),
						Style:      bridge.StyleGraphHorizontalBar,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(stacked-bar)",
						API:        api("/stats/planet/stacked-bar", "get"),
						Style:      bridge.StyleGraphStackedBar,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(horizontal-stacked-bar)",
						API:        api("/stats/planet/horizontal-stacked-bar", "get"),
						Style:      bridge.StyleGraphHorizontalStackedBar,
						Options:    defaultOptions,
						Pagination: false,
					},
					{
						Name:       "Planet(stacked-area)",
						API:        api("/stats/planet/stacked-area", "get"),
						Style:      bridge.StyleGraphStackedArea,
						Options:    defaultOptions,
						Pagination: false,
					},
				},
			},
			// Manage
			{
				ID:      "user",
				Section: bridge.SectionManage,
				Group:   bridge.GroupUser,
				Name:    "ユーザ",
				Components: []*app.Component{
					{
						Name:  "ユーザ",
						API:   api("/user", "get"),
						Style: bridge.StyleTable,
						Query: []*app.Query{
							{
								Key:  "name",
								Type: "string",
							},
						},
						Options:     defaultOptions,
						Pagination:  false,
						TableLabels: []string{"id", "name"},
					},
				},
			},
			{
				ID:      "userblog",
				Section: bridge.SectionManage,
				Group:   bridge.GroupUser,
				Name:    "ユーザブログ",
				Components: []*app.Component{
					{
						Name:        "ユーザーブログ",
						API:         api("/userblog", "get"),
						Style:       bridge.StyleTable,
						Options:     defaultOptions,
						Pagination:  true,
						TableLabels: []string{"id", "user_id"},
					},
				},
			},
			{
				ID:      "userblogentry",
				Section: bridge.SectionManage,
				Group:   bridge.GroupUser,
				Name:    "ユーザブログ記事",
				Components: []*app.Component{
					{
						Name:        "ユーザーブログ記事",
						API:         api("/userblogentry", "get"),
						Style:       bridge.StyleTable,
						Options:     defaultOptions,
						Pagination:  true,
						TableLabels: []string{"id", "user_blog_id"},
					},
				},
			},

			{
				ID:      "blogdesign",
				Name:    "ブログ デザイン",
				Section: bridge.SectionManage,
				Group:   bridge.GroupBlog,
				Components: []*app.Component{
					{
						Name:        "ブログ デザイン",
						API:         api("/blogdesign", "get"),
						Style:       bridge.StyleTable,
						Options:     defaultOptions,
						Pagination:  true,
						TableLabels: []string{"id", "name"},
					},
				},
			},

			{
				ID:      "adminrole",
				Name:    "DMC ユーザー権限",
				Section: bridge.SectionManage,
				Group:   bridge.GroupAdmin,
				Components: []*app.Component{
					{
						Name:        "DMC ユーザー権限",
						API:         api("/adminrole", "get"),
						Style:       bridge.StyleTable,
						Options:     defaultOptions,
						Pagination:  false,
						TableLabels: []string{"role_id"},
					},
				},
			},
			{
				ID:      "adminuser",
				Name:    "DMC ユーザー",
				Section: bridge.SectionManage,
				Group:   bridge.GroupAdmin,
				Components: []*app.Component{
					{
						Name:        "DMC ユーザー権限",
						API:         api("/adminuser", "get"),
						Style:       bridge.StyleTable,
						Options:     defaultOptions,
						Pagination:  false,
						TableLabels: []string{"email", "role_id"},
					},
				},
			},
			{
				ID:      "auditlog",
				Name:    "DMC 監査ログ",
				Section: bridge.SectionManage,
				Group:   bridge.GroupAdmin,
				Components: []*app.Component{
					{
						Name:        "DMC 監査ログ",
						API:         api("/auditlog", "get"),
						Style:       bridge.StyleTable,
						Options:     defaultOptions,
						Pagination:  true,
						TableLabels: []string{"created_at", "request_uri", "request_method"},
					},
				},
			},
		},
	}

	cl := ctx.Context.Value(bridge.JwtClaims)
	if cl != nil {
		// JWTclaimsからRoleを取り出す
		var roles map[string][]string
		claims := cl.(jwtgo.MapClaims)
		json.Unmarshal([]byte(claims["roles"].(string)), &roles)
		res.Pages = apiFilter(res.Pages, roles)
	}

	return ctx.OK(res)
}

File: example-node/api/controllers/dmc.js
const compact = require('mout/array/compact');
const isEmpty = require('mout/lang/isEmpty');
const dmclib = require('node-dmclib');

const constant = require('../../shared/constant');
const helperAdminRole = dmclib.adminRole.helper;

const defaultOptions = {
  key: 'key',
  value: 'value',
};

const api = (path, method) => {
  return {path, method};
};

/**
 * GET: /dmc
 */
const show = (req, res) => {
  const result = {
    name: 'DMC Example Project for Node.js',
    color: 'white',
    thumbnail: 'https://avatars3.githubusercontent.com/u/23251378?v=3&s=200',
    tags: ['develop', 'dmc', 'example'],
    pages: [
      // QuickView
      {
        id: 'quickview',
        name: 'クイックビュー',
        section: constant.DMC_SECTION_DASHBOARD,
        group: constant.GROUP_EMPTY,
        components: [
          {
            name: 'DAU',
            api: api('/stats/dau', 'get'),
            style: constant.DMC_STYLE_NUMBER,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'MAU',
            api: api('/stats/mau', 'get'),
            style: constant.DMC_STYLE_NUMBER,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(bar)',
            api: api('/stats/planet/bar', 'get'),
            style: constant.DMC_STYLE_GRAPH_BAR,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(scatterplot)',
            api: api('/stats/planet/scatterplot', 'get'),
            style: constant.DMC_STYLE_GRAPH_SCATTERPLOT,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(line)',
            api: api('/stats/planet/line', 'get'),
            style: constant.DMC_STYLE_GRAPH_LINE,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(horizontal-bar)',
            api: api('/stats/planet/horizontal-bar', 'get'),
            style: constant.DMC_STYLE_GRAPH_HORIZONTAL_BAR,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(stacked-bar)',
            api: api('/stats/planet/stacked-bar', 'get'),
            style: constant.DMC_STYLE_GRAPH_STACKED_BAR,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(horizontal-stacked-bar)',
            api: api('/stats/planet/horizontal-stacked-bar', 'get'),
            style: constant.DMC_STYLE_GRAPH_HORIZONTAL_STACKED_BAR,
            options: [defaultOptions],
            pagination: false,
          },
          {
            name: 'Planet(stacked-area)',
            api: api('/stats/planet/stacked-area', 'get'),
            style: constant.DMC_STYLE_GRAPH_STACKED_AREA,
            options: [defaultOptions],
            pagination: false,
          },
        ],
      },
      // Manage
      {
        id: 'user',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_USER,
        name: 'ユーザ',
        components: [
          {
            name: 'ユーザ',
            api: api('/user', 'get'),
            style: constant.DMC_STYLE_TABLE,
            query: [
              {
                key: 'name',
                type: 'string',
              },
            ],
            options: [defaultOptions],
            pagination: false,
            table_labels: ['id', 'name'],
          },
        ],
      },
      {
        id: 'userblog',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_USER,
        name: 'ユーザブログ',
        components: [
          {
            name: 'ユーザブログ',
            api: api('/userblog', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['id', 'user_id'],
          },
        ],
      },
      {
        id: 'userblogentry',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_USER,
        name: 'ユーザブログ記事',
        components: [
          {
            name: 'ユーザブログ',
            api: api('/userblogentry', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['id', 'user_blog_id'],
          },
        ],
      },
      {
        id: 'validator',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_TEST,
        name: 'バリデータ検証',
        components: [
          {
            name: 'バリデータ',
            api: api('/validator', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['validation_type'],
          },
        ],
      },
      {
        id: 'nestedarray',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_TEST,
        name: 'ネスト配列',
        components: [
          {
            name: 'ネスト配列',
            api: api('/nestedarray', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['id'],
          },
        ],
      },
      {
        id: 'blogdesign',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_BLOG,
        name: 'ブログ デザイン',
        components: [
          {
            name: 'ブログ デザイン',
            api: api('/blogdesign', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['id', 'name'],
          },
        ],
      },
      // Admin
      {
        id: 'adminrole',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_ADMIN,
        name: 'DMC ユーザ権限',
        components: [
          {
            name: 'DMC ユーザ権限',
            api: api('/adminrole', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: false,
            table_labels: ['role_id'],
          },
        ],
      },
      {
        id: 'adminuser',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_ADMIN,
        name: 'DMC 管理ユーザ',
        components: [
          {
            name: 'DMC 管理ユーザ',
            api: api('/adminuser', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['email', 'role_id'],
          },
        ],
      },
      {
        id: 'auditlog',
        section: constant.DMC_SECTION_MANAGE,
        group: constant.GROUP_ADMIN,
        name: 'DMC 監査ログ',
        components: [
          {
            name: 'DMC 監査ログ',
            api: api('/auditlog', 'get'),
            style: constant.DMC_STYLE_TABLE,
            options: [defaultOptions],
            pagination: true,
            table_labels: ['createdAt', 'request_uri', 'request_method'],
          },
        ],
      },
    ],
  };

  if (!req.swagger.operation.security) {
    // /dmc自体が非認証の場合はそのまま返す
    return res.json(result);
  }

  // 権限がないcomponentを消してから返す
  const roles = req.auth.roles;
  for (let i in result.pages) {
    const page = result.pages[i];
    for (let j in page.components) {
      const component = page.components[j];
      const method = component.api.method;
      const path = component.api.path;
      if (!helperAdminRole.canAccess(path, method, roles)) {
        // 権限がないcomponentを削除
        page.components[j] = null;
      }
    }
    page.components = compact(page.components);
    if (isEmpty(page.components)) {
      // componentが空になった場合はpage自体を削除
      result.pages[i] = null;
    }
  }
  result.pages = compact(result.pages);

  res.json(result);
};

module.exports = {
  'dmc#show': show,
};
